### ISSUE TITLE
avoid multiple materializations on multiple sort

### ISSUE THREAD

**Issue by gab23r** [pola-rs/polars#8410]
### Problem description

If you do `ldf.sort(colum_a).sort(column_b).collect()` it will trigger 2 materializations. This is not the case for `filter` where the optimizer combines the two filters together to avoid unnecessary materialization.

`.sort(colum_a).sort(column_b)` could be computed as `.sort([colum_b, column_a])`

This might looks like a weird opimization because who is going to write two sorts in a row but it actually appears in one of my use case... so why not optimize it ! 

I made a little benchmark just to show what I am saying
```
n_cols = 20
n_rows = 1_000_000
df = pl.LazyFrame(np.random.rand(n_rows, n_cols).round(1))

df_1 = df
for c in df_1.columns:
    df_1 = df_1.sort(c)
df_1 = df_1.collect() #  1.54 s

df_2 = df.sort(df.columns[::-1]).collect() # 173 ms

# df_1 and df_2 are equal
assert df_1.frame_equal(df_2)

```

**Comment by ritchie46** [comment]
Though the output is not guaranteed to be the same I believe. Sorting by column a  and then column b is the same as sorting by column b only.

I do agree we can remove that redundant first sort.

**Comment by StefanBRas** [comment]
> ... Sorting by column a and then column b is the same as sorting by column b only.

I'm not sure this is true if there are ties, right?

```python
import polars as pl
df = pl.LazyFrame({'a': [1,2,3,1,2,3], 'b': [1,2,1,2,1,2], 'c': [1,2,3,4,5,6]})
df.sort('a').sort('b').collect()
shape: (6, 3)
┌─────┬─────┬─────┐
│ a   ┆ b   ┆ c   │
│ --- ┆ --- ┆ --- │
│ i64 ┆ i64 ┆ i64 │
╞═════╪═════╪═════╡
│ 1   ┆ 1   ┆ 1   │
│ 2   ┆ 1   ┆ 5   │
│ 3   ┆ 1   ┆ 3   │
│ 1   ┆ 2   ┆ 4   │
│ 2   ┆ 2   ┆ 2   │
│ 3   ┆ 2   ┆ 6   │
└─────┴─────┴─────┘
df.sort('b').collect()
shape: (6, 3)
┌─────┬─────┬─────┐
│ a   ┆ b   ┆ c   │
│ --- ┆ --- ┆ --- │
│ i64 ┆ i64 ┆ i64 │
╞═════╪═════╪═════╡
│ 1   ┆ 1   ┆ 1   │
│ 3   ┆ 1   ┆ 3   │
│ 2   ┆ 1   ┆ 5   │
│ 2   ┆ 2   ┆ 2   │
│ 1   ┆ 2   ┆ 4   │
│ 3   ┆ 2   ┆ 6   │
└─────┴─────┴─────┘
```

But the documentation makes no promises about keeping the order in case of ties, so maybe that's okay.

**Comment by mcrumiller** [comment]
This would be true if `.sort` had a `maintain_order` parameter, but it doesn't. I don't think there is any guarantee that all items within the same sub-group keep the same order.

**Comment by ritchie46** [comment]
No, I also don't think we should. It really constraints parallelism.

**Comment by gab23r** [comment]
In all my tests, `sort` was maintaining the order, and from a user perspective it is logical as there is no `maintain_order` argument. IMO if something is creating some randomness in the result the user should be able to remove this randomness (for exemple in paying the cost of maintain_order = True)

**Comment by gab23r** [comment]
Now that `maintain_order` exists and is set to False by default. We can keep only the last sort which is going to be an even better optimization!

**Comment by gab23r** [comment]
This issue is outdated and the problem is partially fixed (the first N-1 sort are removed in the maintain_order=False case). I can open a specific issue for the maintain_order=True case.
